rfc/rfc-0014-change-tracking.md

```diff
@@ -0,0 +1,46 @@
+
+It is important to note however, that a workload can have multiple inputs and input into several objects in the supply chain all at the same time.  As a result when a workload changes, it may cause several controllers to do work simultaneously.  
+
+It is also important to note that a controller may also do work outside of the work that the supply chain is choreographing.  Kpack images, for example, are often choreographed as part of a supply chain.  But the kpack controller may also do work in response to a base OS image update.  Nothing to do with the supply chain but impacting it none-the-less.  As a result the supply chain may be triggered part way through by this input.
+Problem
+Whilst, on the one hand, the behavior described above is virtuous and generally beneficial to automated outer loop workflows.  On the other hand, this behavior can be problematic for more user-centric, imperative workflows, such as those often found in the inner loop.  Those of debugging and live update.  
+
+Providing a real world example.  Given a supply chain that has several inputs, let’s say source code image and a debug flag.  When a developer applies a workload after changing their source code and turning on debug.  Several services in the supply chain may trigger at the same time and these inputs will traverse through the supply chain at different times.  But the developer, initiating the debug session will want to wait for that specific source code change to arrive “in cluster” in an image prepped for debugging before attaching their debugger.  So, it is important to know when both inputs have fully traversed the supply chain.
+
+## Possible Solutions
+As the cartographer is already generally aware of workload inputs and outputs (artifacts) and how those artifacts from one component (service) are inputs into another.  It should, therefore, be possible to output, as additional status, the relationship between these artifacts and the service that produced it.
+
+For example:
+
+```
+status:
+   outputs:
+   - type: image
+     value: abc123
+     passed: [image-builder]
+     from:
+     - type: source
+       value: def567
+       passed: [workload, source-cloner, unit-tester]
+```
+
+Here we see the run image `abc123`, produced by the `image-builder` service, and “containing” the source code image `def567`. 
+
+Using the above example again when a workload is applied with the cli, updating both the source image and debug flags, it can then wait for the workload status to be Ready and for an output status showing the last artifact in the supply chain contains the original source image and the debug flag values we expect.
+
+## Concerns/Questions
+How much does a consumer of this output status need to understand about supply chains?  Does a consumer have to “know” what the last artifact produced by a supply chain is?  And how?  
```

Comment:
I'd advocate to let the customer know when its artifact didn't reach the desired state. That means you need to have at least three states e.g. (Pending, Finished or Failed). Otherwise it is difficult to understand what in a coreography went wrong or whether the chain is still under reconciliation.

Reply:
> when its artifact didn't reach the desired state

The artifact is the result of a successful state change. For example, let us say a `ClusterSupplyChain` has a `ClusterSourceTemplate` which templates a FluxCD GitRepository (as in the [examples](https://github.com/vmware-tanzu/cartographer/tree/main/examples/source-to-knative-service)). The artifact of this template is some SHA. At the moment that the artifact, the SHA, is known to Cartographer, the template has succeeded in reaching the desired state. It is impossible to decorate _the artifact_ with any status other than `Successful`/`Finished`.

That said, thinking about UX to enable devs and ops to understand what resources in the SupplyChain are in a pending or failed state is valuable.

Reply:
@waciumawanjohi  Yes, that is right. However there is a lack of visibility during the 20 min or more that could take a pipeline take to product an artifact.